### DISCUSSION TITLE
Why won't getServerSideProps fetch from the backend?

### DISCUSSION THREAD

**Discussion by ChrisB007** [vercel/next.js#29100] [Help] [answered]
Hello, and thanks in advance for your guidance.  So, I am trying to fetch data from my backend using 'getServerSideProps' method. However, when I tried running it I get "undefined", and no further error was logged on the console. Can you tell me where I went wrong here - my code is simplified below

`// "/pages/regularfile.jsx"

import { getSession } from "next-auth/client";

export default function DataContent({ session, project }) {
  console.log({ project }); //Returns undefined :(
  return (
    <>
      <ul>
        <li>{/* Map through the project data */}</li>
      </ul>
    </>
  );
}

export async function getServerSideProps(context) {
  const session = await getSession(context);
  const project = await fetch(
    "http://localhost:3000/api/projects"
  ).then((res) => res.json());
  console.log(project);

  return {
    props: {
      session,
      project
    }
  };
}`

**Comment by wilsonneto-swe** [comment, marked as answer]
Hi man, first an advice, use the code block button here on the textbox to paste code, this will help people to help you, e.g.:
```
// beautiful code here ...
```

Well, about your questions, we have some issues with your code, the correct way to do what you need will be:
(You can't use await and ".then(...)" in the same statement, you need to choose one way to run async code, they do the same thing...)

```
export async function getServerSideProps(context) {
  const session = await getSession(context);

  const res = await fetch("http://localhost:3000/api/projects");
  const projects = await res.json(); 
  
  console.log(project); 

  return {
    props: {
      session,
      project,
    },
  };
}

```

I hope it helps you.
All the best man!

**Reply by ChrisB007** [reply]
Hello @wilsonneto-dev, thanks for the response, and even more thanks for the pointers. I have made changes in my code regarding using "async and .then()"  that you mentioned. However, it's still returning 'undefined' even with the changes - I literally copied and pasted the above into my code. Any more pointers please? This is a link to my working repo : [Nav](https://github.com/ChrisB007/nextFinalPublicTrades/blob/main/components/Navbar.js)

**Reply by wwwdepot** [reply]
`getServerSideProps` only works for [pages](https://nextjs.org/docs/basic-features/pages).

**Reply by ChrisB007** [reply]
Ohhhhhh, wow...Thank you sir. I just learned something new today. Thanks again

**Reply by wilsonneto-swe** [reply]
Wooow man... great, I really forgot to mention this, but this shows that would be a good idea to take some time to read the documentation, the next.js docs are so small and direct, you can view all the docs in minutes... Good luck and success!

**Reply by ChrisB007** [reply]
Thanks

**Reply by phongbach** [reply]
I thought because we cannot call internal API in getServerSideProps, we should move the logic of the api into the function getServerSideProps. I know we should only call API with the full path. However, if we just call our internal API like your previous answer with the full path, is there any down side? The documentation doesnt mention a workaround beside moving the API logic.
